### PR TITLE
hunter: init at 1.3.5

### PR DESCRIPTION
**Pull request by fufexan**
###### Motivation for this change

Add the hunter file manager. Builds on the work done in [99898](https://github.com/NixOS/nixpkgs/pull/99898), since @magnetophon isn't interested in maintaining this package anymore.

The maintainer of the project has disappeared, but there may be folks who decide to fork and continue its development. For this exact reason I have picked a fork as the repo, since it includes some fixes and allows building the package without Rust Nightly.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by legendofmiracles** [comment]
Result of `nixpkgs-review pr 120084` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>hunter</li>
  </ul>
</details>

**Comment by magnetophon** [comment]
@fufexan Thanks!
Cool that there is a more maintained fork now, and that it builds on regular rust.
As far as I can tell, it only has 5 commits on top of upstream.
Let's see where it goes.

Unfortunately, when trying to build it, I get:
```
nixpkgs-review pr 120084
$ git -c fetch.prune=false fetch --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/120084/head:refs/nixpkgs-review/1
$ git worktree add /home/bart/.cache/nixpkgs-review/pr-120084-1/nixpkgs b14062b75c4e8ef4dd4110282f7105be87f681d7
Preparing worktree (detached HEAD b14062b75c4)
Updating files: 100% (25504/25504), done.
HEAD is now at b14062b75c4 zfsUnstable: 2.0.4 -> 2.1.0-rc3
$ git merge --no-commit 0924d47a04689b2d061e8cef02ff2fc4c0037eb4
Automatic merge went well; stopped before committing as requested
$ nix --experimental-features nix-command build --no-link --keep-going --option build-use-sandbox relaxed -f /home/bart/.cache/nixpkgs-review/pr-120084-1/build.nix
building '/nix/store/6kpf15hyr4i6a94xfpxkjrdwyh1n39mb-hunter-1.3.5.drv' on 'ssh://nixBuild@builder'...
[1/0/2 built] building hunter-1.3.5 on ssh://nixBuild@builder: error: build failed
error: build of '/nix/store/6kpf15hyr4i6a94xfpxkjrdwyh1n39mb-hunter-1.3.5.drv' on 'ssh://nixBuild@builder' failed: builder for '/nix/store/6kpf15hyr4i6a94xfpxkjrdwyh1n39mb-hunter-1.3.5.drv' failed with exit code 101
builder for '/nix/store/6kpf15hyr4i6a94xfpxkjrdwyh1n39mb-hunter-1.3.5.drv' failed with exit code 1; last 10 log lines:
     Compiling pin-project-lite v0.2.4
     Compiling getrandom v0.1.16
     Compiling memchr v2.3.4
  error: could not compile `version-compare`

  Caused by:
    process didn't exit successfully: `rustc --crate-name version_compare --edition=2018 /build/hunter-1.3.5-vendor.tar.gz/version-compare/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debug-assertions=off --cfg 'feature="default"' -C metadata=8d0e8559a6e5325f -C extra-filename=-8d0e8559a6e5325f --out-dir /build/source/target/release/deps -C linker=/nix/store/xicw4xz0fnd9kmbdlnj6kajd4wz5im7l-gcc-wrapper-10.2.0/bin/cc -L dependency=/build/source/target/release/deps --cap-lints allow` (signal: 7, SIGBUS: access to undefined memory)
  warning: build failed, waiting for other jobs to finish...
  error: build failed
  builder for '/nix/store/6kpf15hyr4i6a94xfpxkjrdwyh1n39mb-hunter-1.3.5.drv' failed with exit code 101
cannot build derivation '/nix/store/xlryyza6mhmx5zv0jbq0fhi09qy30qwq-env.drv': 1 dependencies couldn't be built
error: build of '/nix/store/xlryyza6mhmx5zv0jbq0fhi09qy30qwq-env.drv' failed

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/120084
```

Any ideas?

**Comment by fufexan** [comment]
@magnetophon are you building on `x86_64-linux`? This is my output:
```
mihai@kiiro ~code/git/nixpkgs % nixpkgs-review pr 120084                                                                                                                       (master)nixpkgs
$ git -c fetch.prune=false fetch --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/120084/head:refs/nixpkgs-review/1
$ git worktree add /home/mihai/.cache/nixpkgs-review/pr-120084-2/nixpkgs b14062b75c4e8ef4dd4110282f7105be87f681d7
Preparing worktree (detached HEAD b14062b75c4)
Updating files: 100% (25504/25504), done.
HEAD is now at b14062b75c4 zfsUnstable: 2.0.4 -> 2.1.0-rc3
$ git merge --no-commit 0924d47a04689b2d061e8cef02ff2fc4c0037eb4
Automatic merge went well; stopped before committing as requested
$ nix --experimental-features nix-command build --no-link --keep-going --option build-use-sandbox relaxed -f /home/mihai/.cache/nixpkgs-review/pr-120084-2/build.nix

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/120084

1 package built:
hunter
```

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 120084` at 0924d47a run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/07b9bc4d57b573e2347a10ada8f7554f">hunter</a></li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement.

**Comment by magnetophon** [comment]
@fufexan Yes, 64 bit linux.
It's on a remote build machine that I just finished and haven't tested much yet, so I guess that's the issue?

Odd, @r-rmcgibbo got a different error, at first glance.

**Comment by fufexan** [comment]
@magnetophon What I can gather from your output, specifically this
```
(signal: 7, SIGBUS: access to undefined memory)
```
is that the command has some kind of problem with memory. Though I'm not sure what specifically. People with more experience in Rust may be able to help us. Or it may be like you said, a problem with your build machine. Try building on another computer if you can.

@r-rmcgibbo It looks like ofborg also failed to build on `aarch64-linux`. It will be resolved upstream (hopefully).

**Comment by magnetophon** [comment]
I just found out that zfs has detected an error on ``/nix/store/8is7s1sgnn9l1kgh2mzirw4b1pbd5r33-rustc-1.51.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-a640413d30ec809b.rlib``, so that is probably the problem, right?

I'll go look up how to force a rebuild.

**Comment by OPNA2608** [comment]
Build fails on x86_64-darwin, you need to add `Security` and `CoreServices`.

**Comment by fufexan** [comment]
@OPNA2608 I hope it's better now. If not, please correct me.

File: pkgs/applications/misc/hunter/default.nix [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  nativeBuildInputs = [ makeWrapper pkg-config ];
+  buildInputs = [
+    glib
+    gst_all_1.gstreamer
+    gst_all_1.gst-plugins-base
+    gst_all_1.gst-plugins-good
+    gst_all_1.gst-plugins-ugly
+    gst_all_1.gst-plugins-bad
+  ] ++ lib.optionals stdenv.isDarwin [ IOKit ];
```
Comment (OPNA2608) [inline pkgs/applications/misc/hunter/default.nix:35]: You forgot to add the packages to buildInputs.

```suggestion
  ] ++ lib.optionals stdenv.isDarwin [ CoreServices IOKit Security ];
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]

**Comment by magnetophon** [comment]
I built it using ``nixpkgs-review pr --build-args="--builders '' --max-jobs 8 " 120084`` and now all is fine.
It also runs.

**Comment by legendofmiracles** [comment]
Hm? How did it compile for me?

> Result of `nixpkgs-review pr 120084` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 1 package built:
> 
>     * hunter
Despite there not being any new commits 🤔

**Comment by OPNA2608** [comment]
Compiles & launches on x86_64-darwin now.

`hunter`
Screenshot_macOS_2021-04-21_21:53:06

I think the `of length 0'` line in the bottom right that was overwritten with huge blank blocks is an error message? It basically just plays a loading animation (3 `.` in front of the text cursor's position).

**Comment by fufexan** [comment]
> Hm? How did it compile for me?
> 
> > Result of `nixpkgs-review pr 120084` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> > 1 package built:
> > ```
> > * hunter
> > ```
> 
> Despite there not being any changes in between thinking

@legendofmiracles magnetophon had issues because of his build machine. OPNA2608 had issues because he's on darwin, and I forgot to add some deps. For r-rmcgibbo it didn't compile because they built it on aarch64 which is currently not supported by hunter. All of these issues had nothing to do with 64bit linux, that's why it built for you from the first try.

If I had any other machines rather than `x86_64-linux`, these issues would have been fixed already.

@OPNA2608 that's weird, I haven't had that happen. Though it's good to know that it now builds on darwin too. :slightly_smiling_face:

**Comment by legendofmiracles** [comment]
Glad we could test it all possible operating systems and architectures 😄

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -14909,6 +14909,8 @@ in
 
   hunspellWithDicts = dicts: callPackage ../development/libraries/hunspell/wrapper.nix { inherit dicts; };
 
+  hunter = callPackage ../applications/misc/hunter { inherit (darwin.apple_sdk.frameworks) CoreServices IOKit Security; };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:14912]:
```suggestion
  hunter = callPackage ../applications/misc/hunter {
    inherit (darwin.apple_sdk.frameworks) CoreServices IOKit Security;
  };
```

File: pkgs/applications/misc/hunter/default.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  '';
+
+  cargoSha256 = "sha256-9DvEuT7Ht0p9vvlQqT/pKL2XuIkJNVAA3KSD+D0Ap38=";
+
+  meta = with lib; {
+    description = "The fastest file manager in the galaxy!";
+    homepage = "https://github.com/rabite0/hunter";
+    license = licenses.wtfpl;
+    maintainers = with maintainers; [ fufexan ];
+    platforms = platforms.unix;
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/hunter/default.nix:50]:
```suggestion
```

set by buildRustPackage to the correct subset of unix.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/hunter/default.nix [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "hunter";
+  version = "1.3.5";
+
+  src = fetchFromGitHub {
+    owner = "06kellyjac";
+    repo = "hunter";
+    rev = "2484f0db580bed1972fd5000e1e949a4082d2f01";
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/hunter/default.nix:21]:
```suggestion
    rev = "v${version }";
```

File: pkgs/applications/misc/hunter/default.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+, IOKit
+, Security
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "hunter";
+  version = "1.3.5";
+
+  src = fetchFromGitHub {
+    owner = "06kellyjac";
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/hunter/default.nix:19]:
```suggestion
    owner = "rabite0";
```

File: pkgs/applications/misc/hunter/default.nix (hunk 4 above)
Comment (fufexan) [reply pkgs/applications/misc/hunter/default.nix:21]: Reverting the commit to the original version means I have to add a rust overlay. The program only worked with nightly rust before.

Do you still want me to do that?

Comment (SuperSandro2000) [reply pkgs/applications/misc/hunter/default.nix:21]: Please fetch the release and the two commits as patches instead of fetching the fork branch.

Comment (fufexan) [reply pkgs/applications/misc/hunter/default.nix:21]: The commits ((this)[https://github.com/06kellyjac/hunter/commit/a5943578e1ee679c8bc51b0e686c6dddcf74da2a] and (this)[https://github.com/06kellyjac/hunter/commit/1b8de9248312878358afaf1dac569ebbccc4321a]) don't get patched as expected *on top* of the 1.3.5 version. They're based on the latest commit, which is more than 70 commits ahead of 1.3.5. I guess I should switch the version to the latest commit as well then, and change the version to `2020-05-25-unstable`.
What do you think @SuperSandro2000 ?

Comment (SuperSandro2000) [reply pkgs/applications/misc/hunter/default.nix:21]: Sure. If you think the software works better at this commit and does not contain any new bugs.

Comment (fufexan) [reply pkgs/applications/misc/hunter/default.nix:21]: I found something interesing: even if `owner = "rabite0"`, specifying the latest commit from 06kellyjac's fork results in fetching that version of the repo and it builds just fine.
Should I still have the extra commits as patches or is it fine to let the `rev` be the latest commit from that fork?

Comment (SuperSandro2000) [reply pkgs/applications/misc/hunter/default.nix:21]: * ``020-05-25-unstable``

> I found something interesing: even if owner = "rabite0", specifying the latest commit from 06kellyjac's fork results in fetching that version of the repo and it builds just fine.

GitHub thought this is a good "feature" but it is just a security nightmare. Please fetch the commit from the repo where it belongs/originates to.

**Comment by fufexan** [comment]
I've tried adding patches as @SuperSandro2000 suggested, however the build fails to use `cargoSha256` and complains about it being outdated. I can't get it to give me a "newer" one. I'm posting the derivation I got so far. If anyone could help me I'd be grateful.
```nix
{ lib
, stdenv
, pkg-config
, rustPlatform
, fetchFromGitHub
, fetchpatch
, makeWrapper
, glib
, gst_all_1
, CoreServices
, IOKit
, Security
}:

rustPlatform.buildRustPackage rec {
  pname = "hunter";
  version = "2020-05-25-unstable";

  src = fetchFromGitHub {
    owner = "rabite0";
    repo = "hunter";
    rev = "355d9a3101f6d8dc375807de79e368602f1cb87d";
    sha256 = "sha256-R2wNkG8bFP7X2pdlebHK6GD15qmD/zD3L0MwVthvzzQ=";
  };

  patches = [
    (fetchpatch {
      name = "remove-dependencies-on-rust-nightly";
      url =
        "https://github.com/06kellyjac/hunter/commit/a5943578e1ee679c8bc51b0e686c6dddcf74da2a.diff";
      sha256 = "sha256-eOwBFfW5m8tPnu+whWY/53X9CaqiVj2WRr25G+Yy7qE=";
    })
    (fetchpatch {
      name = "fix-accessing-core-when-moved-with-another-clone";
      url =
        "https://github.com/06kellyjac/hunter/commit/2e95cc567c751263f8c318399f3c5bb01d36962a.diff";
      sha256 = "sha256-yTzIXUw5qEaR2QZHwydg0abyZVXfK6fhJLVHBI7EAro=";
    })
    (fetchpatch {
      name = "chore-cargo-update";
      url =
        "https://github.com/06kellyjac/hunter/commit/b0be49a82191a4420b6900737901a71140433efd.diff";
      sha256 = "sha256-ctxoDwyIJgEhMbMUfrjCTy2SeMUQqMi971szrqEOJeg=";
    })
    (fetchpatch {
      name = "chore-cargo-upgrade-+-cargo-update";
      url =
        "https://github.com/06kellyjac/hunter/commit/1b8de9248312878358afaf1dac569ebbccc4321a.diff";
      sha256 = "sha256-+4DZ8SaKwKNmr2SEgJJ7KZBIctnYFMQFKgG+yCkbUv0=";
    })
    (fetchpatch {
      name = "fix-resolve-breaking-changes-from-package-updates";
      url =
        "https://github.com/06kellyjac/hunter/commit/2484f0db580bed1972fd5000e1e949a4082d2f01.diff";
      sha256 = "sha256-K+WUxEr1eE68XejStj/JwQpMHlhkiOw6PmiSr1GO0kc=";
    })
  ];

  RUSTC_BOOTSTRAP = 1;

  nativeBuildInputs = [ makeWrapper pkg-config ];
  buildInputs = [
    glib
    gst_all_1.gstreamer
    gst_all_1.gst-plugins-base
    gst_all_1.gst-plugins-good
    gst_all_1.gst-plugins-ugly
    gst_all_1.gst-plugins-bad
  ] ++ lib.optionals stdenv.isDarwin [ CoreServices IOKit Security ];

  cargoBuildFlags = [ "--no-default-features" "--features=img,video" ];

  postInstall = ''
    wrapProgram $out/bin/hunter --prefix GST_PLUGIN_SYSTEM_PATH_1_0 : "$GST_PLUGIN_SYSTEM_PATH_1_0"
  '';

  cargoSha256 = "sha256-+1M22yQ8g9pKDYf5Zm9ae6oawoKB7kqFjykKbDeGDg8=";

  meta = with lib; {
    description = "The fastest file manager in the galaxy!";
    homepage = "https://github.com/rabite0/hunter";
    license = licenses.wtfpl;
    maintainers = with maintainers; [ fufexan ];
  };
}
```

File: pkgs/applications/misc/hunter/default.nix [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+  nativeBuildInputs = [ makeWrapper pkg-config ];
+  buildInputs = [
+    glib
+    gst_all_1.gstreamer
+    gst_all_1.gst-plugins-base
+    gst_all_1.gst-plugins-good
+    gst_all_1.gst-plugins-ugly
+    gst_all_1.gst-plugins-bad
+  ] ++ lib.optionals stdenv.isDarwin [ CoreServices IOKit Security ];
```
Comment (OPNA2608) [inline pkgs/applications/misc/hunter/default.nix:61]:
```suggestion
  ] ++ (with gst_all_1; [
    gstreamer
    gst-plugins-base
    gst-plugins-good
    gst-plugins-ugly
    gst-plugins-bad
  ]) ++ lib.optionals stdenv.isDarwin [ CoreServices IOKit Security ];
```

File: pkgs/applications/misc/hunter/default.nix [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  '';
+
+  cargoSha256 = "sha256-Bd/gilebxC4H+/1A41OSSfWBlHcSczsFcU2b+USnI74=";
+
+  meta = with lib; {
+    description = "The fastest file manager in the galaxy!";
+    homepage = "https://github.com/rabite0/hunter";
+    license = licenses.wtfpl;
+    maintainers = with maintainers; [ fufexan ];
+  };
```
Comment (OPNA2608) [inline pkgs/applications/misc/hunter/default.nix:76]: Doesn't seem to build on ARM, [see OfBorg logs](https://logs.nix.ci/?key=nixos/nixpkgs.120084&attempt_id=77affb68-aed1-4d4a-b19d-557f3eacb5d3).
```suggestion
    # error[E0308]: mismatched types
    # --> src/files.rs:502:62
    # expected raw pointer `*const u8`, found raw pointer `*const i8`
    broken = stdenv.isAarch64;
  };
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Result of `nixpkgs-review pr 120084` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>hunter</li>
  </ul>
</details>

---

Binary launches but I have no clue how to navigate it.

**Comment by fufexan** [comment]
> Binary launches but I have no clue how to navigate it.

Arrow keys / hjkl to get around dirs, `/` to search. All the commands are found in the project's readme.

**Comment by fufexan** [comment]
Any news on this? Should I run ofborg again?

**Comment by SuperSandro2000** [comment]
@ofborg eval

File: pkgs/applications/misc/hunter/default.nix [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    (fetchpatch {
+      name = "chore-cargo-update";
+      url =
+        "https://github.com/06kellyjac/hunter/commit/b0be49a82191a4420b6900737901a71140433efd.diff";
+      sha256 = "sha256-ctxoDwyIJgEhMbMUfrjCTy2SeMUQqMi971szrqEOJeg=";
+    })
+    (fetchpatch {
+      name = "chore-cargo-upgrade-+-cargo-update";
+      url =
+        "https://github.com/06kellyjac/hunter/commit/1b8de9248312878358afaf1dac569ebbccc4321a.diff";
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/hunter/default.nix:46]:
```suggestion
      url = "https://github.com/06kellyjac/hunter/commit/a5943578e1ee679c8bc51b0e686c6dddcf74da2a.diff";
      sha256 = "sha256-eOwBFfW5m8tPnu+whWY/53X9CaqiVj2WRr25G+Yy7qE=";
    })
    (fetchpatch {
      name = "fix-accessing-core-when-moved-with-another-clone";
      url = "https://github.com/06kellyjac/hunter/commit/2e95cc567c751263f8c318399f3c5bb01d36962a.diff";
      sha256 = "sha256-yTzIXUw5qEaR2QZHwydg0abyZVXfK6fhJLVHBI7EAro=";
    })
    (fetchpatch {
      name = "fix-resolve-breaking-changes-from-package-updates";
      url = "https://github.com/06kellyjac/hunter/commit/2484f0db580bed1972fd5000e1e949a4082d2f01.diff";
      sha256 = "sha256-K+WUxEr1eE68XejStj/JwQpMHlhkiOw6PmiSr1GO0kc=";
    })
  ];

  cargoPatches = [
    (fetchpatch {
      name = "chore-cargo-update";
      url = "https://github.com/06kellyjac/hunter/commit/b0be49a82191a4420b6900737901a71140433efd.diff";
      sha256 = "sha256-ctxoDwyIJgEhMbMUfrjCTy2SeMUQqMi971szrqEOJeg=";
    })
    (fetchpatch {
      name = "chore-cargo-upgrade-+-cargo-update";
      url = "https://github.com/06kellyjac/hunter/commit/1b8de9248312878358afaf1dac569ebbccc4321a.diff";
```

**Pull request merged**